File: bakamc-folia/src/main/kotlin/cn/bakamc/folia/db/table/SpecialItem.kt
package cn.bakamc.folia.db.table

import cn.bakamc.folia.util.logger
import com.mojang.serialization.JsonOps
import moe.forpleuvoir.nebula.serialization.gson.parseToJsonElement
import moe.forpleuvoir.nebula.serialization.gson.toJsonStr
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.item.ItemStack
import org.ktorm.database.Database
import org.ktorm.entity.Entity
import org.ktorm.entity.sequenceOf
import org.ktorm.schema.Table
import org.ktorm.schema.blob
import org.ktorm.schema.varchar

open class SpecialItems(alias: String?) : Table<SpecialItem>("special_item", alias) {
    companion object : SpecialItems(null)

    override fun aliased(alias: String): Table<SpecialItem> = SpecialItems(alias)

    val key = varchar("key").primaryKey().bindTo { it.key }

    val id = varchar("id").bindTo { it.id }

    val itemData = blob("item_data").bindTo { it.itemData }

}

interface SpecialItem : Entity<SpecialItem> {
    companion object : Entity.Factory<SpecialItem>()

    var key: String

    var id: String

    var itemData: ByteArray

}

fun ItemStack.toSpecialItem(key: String): SpecialItem {
    return SpecialItem {
        this.key = key
        this.id = nameSpace
        val result = ItemStack.SINGLE_ITEM_CODEC.encodeStart(JsonOps.COMPRESSED, this@toSpecialItem).result()
        this.itemData = ByteArray(0)
        result.ifPresent { this.itemData = it.toJsonStr().toByteArray(Charsets.UTF_8) }
    }
}

val Database.specialItems get() = this.sequenceOf(SpecialItems)

val ItemStack.nameSpace: String get() = BuiltInRegistries.ITEM.getKey(this.item).toString()

fun SpecialItem.toItemStack(count: Int): ItemStack? {
    runCatching {
        val item = ItemStack.SINGLE_ITEM_CODEC.decode(JsonOps.COMPRESSED, itemData.toString(Charsets.UTF_8).parseToJsonElement).result().get().first
        return item.apply {
            this.count = count
        }
    }.onFailure {
        logger.info("物品转换异常")
        it.printStackTrace()
    }
    return null
}

fun SpecialItem.isMatch(item: ItemStack): Boolean {
    runCatching {
        return ItemStack.isSameItemSameComponents(item, toItemStack(item.count)!!)
    }.onFailure {
        logger.info("物品匹配异常")
        it.printStackTrace()
    }
    return false
}

File: bakamc-folia/src/main/scala/cn/bakamc/folia/database/table/SpecialItem.scala
package cn.bakamc.folia.database.table

import cn.bakamc.folia.BakaMC
import cn.bakamc.folia.util.text.GSON
import com.google.gson.JsonParser
import com.mojang.serialization.JsonOps
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.world.item.{Item as MCItem, ItemStack as MCItemStack}
import org.bukkit.craftbukkit.CraftRegistry
import org.bukkit.craftbukkit.inventory.CraftItemStack
import org.bukkit.inventory.ItemStack
import slick.jdbc.H2Profile.Table
import slick.jdbc.MySQLProfile.api.*
import slick.lifted.{ProvenShape, Tag}

import java.nio.charset.StandardCharsets

case class SpecialItem(
  id: String,
  namespace: String,
  itemData: Array[Byte]
) {

  private[table] lazy val stackCache: Option[MCItemStack] = {
    try {
      val registry = CraftRegistry.getMinecraftRegistry
      val item = MCItemStack.SINGLE_ITEM_CODEC.decode(
        registry.createSerializationContext(JsonOps.COMPRESSED),
        JsonParser.parseString(String(itemData, StandardCharsets.UTF_8))
      ).result.get.getFirst
      Some(item)
    } catch {
      case e: Throwable =>
        BakaMC.logger.error("物品转换异常", e)
        None
    }
  }

}

object SpecialItem {

  given Conversion[MCItemStack, MCItem] = _.getItem

  def apply(id: String, nameSpace: String, itemData: Array[Byte]): SpecialItem = new SpecialItem(id, nameSpace, itemData)

  def apply(itemStack: ItemStack | MCItemStack): SpecialItem = {
    SpecialItem(getItemName(itemStack), itemStack)
  }

  def apply(id: String, itemStack: ItemStack | MCItemStack): SpecialItem = {
    val ops = CraftRegistry.getMinecraftRegistry.createSerializationContext(JsonOps.COMPRESSED)
    val stack = itemStack match {
      case item: MCItemStack => item
      case item: ItemStack => CraftItemStack.asNMSCopy(item)
    }
    val data = GSON.toJson(MCItemStack.SINGLE_ITEM_CODEC.encodeStart(ops, stack).result().get()).getBytes(StandardCharsets.UTF_8)
    SpecialItem(id, getItemNamespace(stack), data)
  }

  def getItemName(item: MCItemStack | ItemStack): String = {
    item match {
      case item: MCItemStack =>
        item.getItemName.getString
      case item: ItemStack =>
        CraftItemStack.asNMSCopy(item).getItemName.getString
    }
  }

  def getItemNamespace(item: MCItem | ItemStack): String = {
    item match {
      case item: MCItem =>
        BuiltInRegistries.ITEM.getKey(item).toString
      case item: ItemStack =>
        BuiltInRegistries.ITEM.getKey(CraftItemStack.asNMSCopy(item)).toString
    }
  }

  extension (self: SpecialItem) {

    def getVanillaItemStack(count: Int = 1): Option[MCItemStack] = {
      self.stackCache match {
        case Some(item) => Some(item.copyWithCount(count))
        case None => None
      }
    }

    def getItemStack(count: Int = 1): Option[ItemStack] = {
      getVanillaItemStack(count) match {
        case Some(item) => Some(CraftItemStack.asBukkitCopy(item))
        case None => None
      }
    }
  }
}

class SpecialItems(tag: Tag) extends Table[SpecialItem](tag, "special_items") {

  def id = column[String]("id", O.PrimaryKey, O.Length(36))

  def namespace = column[String]("namespace", O.Length(36))

  def itemData = column[Array[Byte]]("item_data")

  override def * : ProvenShape[SpecialItem] = (id, namespace, itemData).mapTo[SpecialItem]
}

val specialItems = TableQuery[SpecialItems]